NF-schema refactor of BUSCO/Omark.

ens-LCampbell opened the pull request:
Take no. 2 😄 

Thanks for making the prerequisite initial merge to Main folks. Im happy we can properly preserve the ownership and credit to @ens-ftricomi, also now the merge seems to be fine  with no major conflicts this time around.  I have done my upmost to preserve the code and the pipeline output. 

**In summary the main changes are as follows:**

- Removed use of all java mysql code and associated plugins (reason for this -> Grab @ class imports are no longer supported in later versions of NXF). Literally overnight it was working this week, then next day this import was causing nxf to crash out. 
- Remaining we have one import of **JsonSlurper()**, which I believe could also be coopted soon too but I need more time to check what the nextflow community recommend. 
 
- Required now to have NXF version 23.10+ (associated with using nf-schema 2.2+)
- Introduced [nf-schema 2.2](https://nextflow-io.github.io/nf-schema/latest/) and pipeline param runtime validation and auto help documentation. 
- Add Nextflow-schema, and samplesheetToList and CSV file validation. 
- A new database meta table module, subworkflow and indexing function (`index_metadata` in utils.nf) to facilitate meta data propagation and flow through. The module query_core_meta is standalone and relies on ensembl-genomio (version set latest v1.6.1) python code accessed via a its [docker container](https://hub.docker.com/r/ensemblorg/ensembl-genomio/tags) and an input query JSON file `meta_filter.json`. Note  obviously nothing set in stone, any feedback welcome we can work to common goal. 

- Refinements to some modules in my attempt to adhere to the slowly developing [Ensembl Nextflow coding guidelines](https://www.ebi.ac.uk/seqdb/confluence/display/EnsGen/Nextflow+coding+guidelines), removal of redundant modules where possible (e.g. BUSCO_GENOME|BUSCO_PROTEIN_LINEAGE  now combined to BUSCO_LINEAGES).
- Added a user helper script to generate the appropriate pipeline run command. This script will aid in ENV setup and checks for appropriate dependencies. However this script was written with a 'ensembl-mod-env' specific setup in mind. Obviously the script isn't required, but rather aids those that chose to use it to create the long NXF run command. 
- Added conda env setup via nextflow for those python specific modules requiring associated dependencies (i.e requirements.txt incorporated via new `environment.yml`). Tested and working as expected. My suggestion is to all users to add associated env variable NXF_CONDA_CACHEDIR (see [here](https://www.nextflow.io/docs/latest/reference/env-vars.html)) to point to a shared filesystem space to aid in reuse of existing conda envs. Can help if needed.
            - How to allow conda env use if you're base environment doesn't have python+deps needed: `nextflow run main.nf .... -profile conda`, or `-profile slurm,conda` 
- Some new run parameters to help different teams run the pipeline. e.g. `--server_set ensrw` to help with mysql host command paths.

Not addressed :
- README (help doc section etc). If you want this update to follow output using `--showHelp --showFull` I can add to this PR. I can produce using nf-core MD/HTML format help documentation so please shout and I will add this. 
- Some additional development might be needed to see if we can remove use of hardcoded PATHs being exposed to the public. E.g. mysql_commands, production_tp_dir etc etc. 

Moving forward:
- We have tested on the plantazoa side extensively. But of course would very much like Genebuild users feedback. @EreboPSilva I believe you might be the man to poke for this 👍 

Given this scope and range of changes to the existing pipeline, my suggestion would be to _before merging_ the PR, create a **tagged release** version. This will ensure all existing code is preserved and available should it be needed.

Commit:
File name changes + updates to these files

Commit:
New configs, schema + nf files added to the pipeline

Commit:
Updates to existing conf files

Commit:
Changes to main workflow and nxf config

Commit:
Updated BUSCO modules

Commit:
Updated ensembl_statistics modules

Commit:
Updated Fetch modules

Commit:
Updated Omark modules

Commit:
Updated Subworkflows

Commit:
Updated utils/misc modules

Commit:
Requirements.txt tweak

Commit:
Refinements after further testing

Commit:
Conda_env testing and refinements

Commit:
Setup helper script tweak

leannehaggerty commented:
Just to note
@JackCurragh and @ens-ftricomi have been added to review and approve.
@leannehaggerty  and @EreboPSilva have been added to test in our envs.

JAlvarezJarreta commented on pipelines/nextflow/conf/genomio.config:
```diff
@@ -0,0 +1,28 @@
+*/
+
+params {
+        genomio_version = 'v1.6.1'
```
```suggestion
    genomio_version = 'v1.6.1'
```

JAlvarezJarreta commented on pipelines/nextflow/modules/busco/busco_core_metakeys.nf:
```diff
@@ -16,47 +15,23 @@ See the License for the specific language governing permissions and
-    
-
-    
+        busco_metakeys_script = file("${workflow.projectDir}/bin/busco_metakeys_patch.py")
```
```suggestion
        def busco_metakeys_script = file("${workflow.projectDir}/bin/busco_metakeys_patch.py")
```

JAlvarezJarreta commented on pipelines/nextflow/modules/busco/busco_dataset.nf:
```diff
@@ -17,23 +16,26 @@ limitations under the License.
-    """
-
-    
+        output = 'orthodb_set.txt'
+        orthodb = ''
```
```suggestion
        def output = 'orthodb_set.txt'
        def orthodb = ''
```

JAlvarezJarreta commented on pipelines/nextflow/modules/busco/busco_lineages.nf:
```diff
@@ -0,0 +1,60 @@
+
+    script:
+        if ( input_busco_mode == 'protein' ) {
+            outdir = 'protein_output'
+            busco_mode = 'proteins'
```
```suggestion
            def outdir = 'protein_output'
            def busco_mode = 'proteins'
```

JAlvarezJarreta commented on pipelines/nextflow/modules/busco/busco_lineages.nf:
```diff
@@ -0,0 +1,60 @@
+            busco_mode = 'proteins'
+        }
+        else if( input_busco_mode == 'genome' ) {
+            outdir = 'genome_output'
+            busco_mode = input_busco_mode
```
```suggestion
            def outdir = 'genome_output'
            def busco_mode = input_busco_mode
```

JAlvarezJarreta commented on pipelines/nextflow/modules/busco/busco_output.nf:
```diff
@@ -16,48 +15,41 @@ See the License for the specific language governing permissions and
-    """
-    sed '/Summarized benchmarking in BUSCO notation for file/d' $summary_file > $summary_name
-    """
+        formated_sci_name = organism_name.replaceAll("\\s", "_")
+        publish_dir_name = formated_sci_name + '/' + insdc_acc + '/' + annotation_source
+        species_lc = formated_sci_name.toLowerCase()
```
```suggestion
        def formated_sci_name = organism_name.replaceAll("\\s", "_")
        def publish_dir_name = formated_sci_name + '/' + insdc_acc + '/' + annotation_source
        def species_lc = formated_sci_name.toLowerCase()
```

JAlvarezJarreta commented on pipelines/nextflow/modules/busco/busco_output.nf:
```diff
@@ -16,48 +15,41 @@ See the License for the specific language governing permissions and
+        species_lc = formated_sci_name.toLowerCase()
+
+        if (dbname=='dummycore') {
+            publish_dir_name = insdc_acc
+            species_lc = 'species=NA'
```
```suggestion
            def publish_dir_name = insdc_acc
            def species_lc = 'species=NA'
```

JAlvarezJarreta commented on pipelines/nextflow/modules/busco/busco_output.nf:
```diff
@@ -16,48 +15,41 @@ See the License for the specific language governing permissions and
+        }
+
+        if (datatype == "genome") {
+            busco_sum = "genome_busco"
```
```suggestion
            def busco_sum = "genome_busco"
```

JAlvarezJarreta commented on pipelines/nextflow/modules/busco/busco_output.nf:
```diff
@@ -16,48 +15,41 @@ See the License for the specific language governing permissions and
+            busco_sum = "genome_busco"
+        }
+        else if (datatype == "protein") {
+            busco_sum = "protein_busco"
```
```suggestion
            def busco_sum = "protein_busco"
```

JAlvarezJarreta commented on pipelines/nextflow/modules/busco/busco_output.nf:
```diff
@@ -16,48 +15,41 @@ See the License for the specific language governing permissions and
+        publish_dir_name = formated_sci_name + '/' + insdc_acc + '/' + annotation_source
+        species_lc = formated_sci_name.toLowerCase()
+
+        if (dbname=='dummycore') {
```
```suggestion
        if (dbname =='dummycore') {
```

JAlvarezJarreta commented on pipelines/nextflow/modules/busco/busco_output.nf:
```diff
@@ -16,48 +15,41 @@ See the License for the specific language governing permissions and
+            busco_sum = "protein_busco"
+        }
+
+        gca_string = insdc_acc.toLowerCase().replaceAll(/\./, "v").replaceAll(/_/, "")
+        summary_name = [species_lc, gca_string, busco_sum, "short_summary.txt"].join("_")
```
```suggestion
        def gca_string = insdc_acc.toLowerCase().replaceAll(/\./, "v").replaceAll(/_/, "")
        def summary_name = [species_lc, gca_string, busco_sum, "short_summary.txt"].join("_")
```

ens-ftricomi replied:
same comment, it can simply be the output from busco and the module for the patches can simply point to that one. the previous name convention was needed fot the public ftp

JAlvarezJarreta commented on pipelines/nextflow/modules/copy_output_to_ensembl_ftp.nf:
```diff
@@ -16,29 +15,25 @@ See the License for the specific language governing permissions and
-    """
-    //sudo -u genebuild rsync -ahvW $summary_file $ftp_stats && rsync -avhc $summary_file $ftp_stats; \
-}
+        statistics_files = file("${params.outDir}/${publish_dir_name}/statistics/*summary.txt")
+        read_me = file("${workflow.projectDir}/../data/README.txt")
+        ftp_stats = "${params.production_ftp_dir}/${publish_dir_name}/statistics"
+        // ftp_path = "${params.production_ftp_dir}/$formated_sci_name"
```
```suggestion
        def statistics_files = file("${params.outDir}/${publish_dir_name}/statistics/*summary.txt")
        def read_me = file("${workflow.projectDir}/../data/README.txt")
        def ftp_stats = "${params.production_ftp_dir}/${publish_dir_name}/statistics"
```

JAlvarezJarreta commented on pipelines/nextflow/modules/copy_output_to_ensembl_ftp.nf:
```diff
@@ -16,29 +15,25 @@ See the License for the specific language governing permissions and
+        sudo -u genebuild chmod 775 ${ftp_stats}/* -R;
+        sudo -u genebuild chgrp ensemblftp ${ftp_stats}/* -R;
+        """
+        //sudo -u genebuild rsync -ahvW $summary_file $ftp_stats && rsync -avhc $summary_file $ftp_stats; \
```
```suggestion
```

JAlvarezJarreta commented on pipelines/nextflow/modules/ensembl_statistics/populate_db.nf:
```diff
@@ -16,20 +15,22 @@ See the License for the specific language governing permissions and
-    """
-}
+        if (params.server_set){
+            host_admin = params.server_set
```
```suggestion
            def host_admin = params.server_set
```

JAlvarezJarreta commented on pipelines/nextflow/modules/ensembl_statistics/populate_db.nf:
```diff
@@ -16,20 +15,22 @@ See the License for the specific language governing permissions and
+            host_admin = params.server_set
+        }
+        else{
+            host_admin = params.mysql_ensadmin
```
```suggestion
            def host_admin = params.mysql_ensadmin
```

JAlvarezJarreta commented on pipelines/nextflow/modules/ensembl_statistics/populate_db.nf:
```diff
@@ -16,20 +15,22 @@ See the License for the specific language governing permissions and
+        if (params.server_set){
+            host_admin = params.server_set
+        }
+        else{
```
```suggestion
        else {
```

JAlvarezJarreta commented on pipelines/nextflow/modules/ensembl_statistics/run_ensembl_meta.nf:
```diff
@@ -16,48 +15,36 @@ See the License for the specific language governing permissions and
+            val(production_name), val(organism_name), val(annotation_source), path("*.sql")
     
+    script:
+        formated_sci_name = organism_name.replaceAll("\\s", "_")
+        publish_dir_name = formated_sci_name + '/' + insdc_acc + '/' + annotation_source
+        core_meta_data_script = file("${params.enscode}/ensembl-genes/src/python/ensembl/genes/metadata/core_meta_data.py")
+        output_dir = file("${params.outDir}/${publish_dir_name}/")
```
```suggestion
        def formated_sci_name = organism_name.replaceAll("\\s", "_")
        def publish_dir_name = formated_sci_name + '/' + insdc_acc + '/' + annotation_source
        def core_meta_data_script = file("${params.enscode}/ensembl-genes/src/python/ensembl/genes/metadata/core_meta_data.py")
        def output_dir = file("${params.outDir}/${publish_dir_name}/")
```

JAlvarezJarreta commented on pipelines/nextflow/modules/ensembl_statistics/run_statistics.nf:
```diff
@@ -16,39 +15,32 @@ See the License for the specific language governing permissions and
-    //gb1-w amphiduros_pacificus_gca949316495v1_core_110_1 <stats_amphiduros_pacificus_gca949316495v1_core_110_1.sql
-    
-
+        formated_sci_name = organism_name.replaceAll("\\s", "_")
+        publish_dir_name = formated_sci_name + '/' + insdc_acc + '/' + annotation_source
+        corestats_outdir = "core_statistics"
+        stats_script = file("${params.enscode}/ensembl-genes/src/python/ensembl/genes/stats/generate_species_homepage_stats.pl")
```
```suggestion
        def formated_sci_name = organism_name.replaceAll("\\s", "_")
        def publish_dir_name = formated_sci_name + '/' + insdc_acc + '/' + annotation_source
        def corestats_outdir = "core_statistics"
        def stats_script = file("${params.enscode}/ensembl-genes/src/python/ensembl/genes/stats/generate_species_homepage_stats.pl")
```

JAlvarezJarreta commented on pipelines/nextflow/modules/fetch_genome.nf:
```diff
@@ -18,23 +17,25 @@ limitations under the License.
-
+    tag "${organism_name}:${insdc_acc}"
+    label 'fetch_file'
+    storeDir "${params.cacheDir}/${insdc_acc}/ncbi_dataset/" // update on protein busco side
```
```suggestion
    storeDir "${params.cacheDir}/${insdc_acc}/ncbi_dataset/"  // update on protein busco side
```

JAlvarezJarreta commented on pipelines/nextflow/modules/fetch_genome.nf:
```diff
@@ -18,23 +17,25 @@ limitations under the License.
+            val(ortho_db), path("*.fna"), emit: genome_fasta
+    
+    script:
+        outfile = "genome_file.zip"
```
```suggestion
        def outfile = "genome_file.zip"
```

JAlvarezJarreta commented on pipelines/nextflow/modules/fetch_genome_core.nf:
```diff
@@ -16,33 +15,33 @@ See the License for the specific language governing permissions and
- 
- """
+    genome_fasta = "genome_toplevel.fa"
+    sequence_dumping_script = file("${params.enscode}/ensembl-analysis/scripts/sequence_dump.pl")
```
```suggestion
    def genome_fasta = "genome_toplevel.fa"
    def sequence_dumping_script = file("${params.enscode}/ensembl-analysis/scripts/sequence_dump.pl")
```

JAlvarezJarreta commented on pipelines/nextflow/modules/fetch_proteins.nf:
```diff
@@ -16,31 +15,32 @@ See the License for the specific language governing permissions and
-        -file $translations_file \
-        ${params.dump_params}
-    """
+        translations_file = production_name +"_translations.fa"
+        dump_translations_script = file("${params.enscode}/ensembl-analysis/scripts/protein/dump_translations.pl")
```
```suggestion
        def translations_file = production_name +"_translations.fa"
        def dump_translations_script = file("${params.enscode}/ensembl-analysis/scripts/protein/dump_translations.pl")
```

ens-ftricomi replied:
sometimes create conflict so I would test it before accepting it

JAlvarezJarreta commented on pipelines/nextflow/modules/omark/omark.nf:
```diff
@@ -16,28 +15,24 @@ See the License for the specific language governing permissions and
-    afterScript "sleep $params.files_latency"  // Needed because of file system latency
-    maxForks 15
+    tag "${organism_name}:${insdc_acc}"
+    container "${params.omark_singularity_path}"
+    publishDir "${params.outDir}/${publish_dir_name}/", mode: 'copy'
```
```suggestion
    publishDir "${params.outDir}/${publish_dir_name}/", mode: 'copy'

    container "${params.omark_singularity_path}"
```

JAlvarezJarreta commented on pipelines/nextflow/modules/omark/omark.nf:
```diff
@@ -16,28 +15,24 @@ See the License for the specific language governing permissions and
-    scientific_name = getMetaValue(db, "species.scientific_name")[0].meta_value.toString().replaceAll("\\s", "_")
-    species=scientific_name.toLowerCase()
-    publish_dir =scientific_name +'/'+gca+'/'+getMetaValue(db, "species.annotation_source")[0].meta_value.toString()
+        formated_sci_name = organism_name.replaceAll("\\s", "_")
+        publish_dir_name = formated_sci_name + '/' + insdc_acc + '/' + annotation_source
```
```suggestion
        def formated_sci_name = organism_name.replaceAll("\\s", "_")
        def publish_dir_name = formated_sci_name + '/' + insdc_acc + '/' + annotation_source
```

JAlvarezJarreta commented on pipelines/nextflow/modules/omark/omark_output.nf:
```diff
@@ -16,30 +15,23 @@ See the License for the specific language governing permissions and
-    
-    """
+        accession_formatted = insdc_acc.toLowerCase().replaceAll(/\./, "v").replaceAll(/_/, "")
+        summary_name = [formated_sci_name.toLowerCase(), accession_formatted, "omark", "proteins_detailed_summary.txt"].join("_")
```
```suggestion
        def accession_formatted = insdc_acc.toLowerCase().replaceAll(/\./, 'v').replaceAll(/_/, '')
        def summary_name = [formated_sci_name.toLowerCase(), accession_formatted, 'omark', 'proteins_detailed_summary.txt'].join('_')
```

JAlvarezJarreta commented on pipelines/nextflow/modules/query_core_meta.nf:
```diff
@@ -0,0 +1,34 @@
+
+    input:
+        val(core)
+        path(meta_keys)
```
```suggestion
        val core
        path meta_keys
```

JAlvarezJarreta commented on pipelines/nextflow/modules/query_core_meta.nf:
```diff
@@ -0,0 +1,34 @@
+        path(meta_keys)
+
+    output:
+        path("coredb_meta.json"), emit: metadata_json
```
```suggestion
        path 'coredb_meta.json', emit: metadata_json
```

JAlvarezJarreta commented on pipelines/nextflow/subworkflows/run_busco.nf:
```diff
@@ -18,24 +17,23 @@ limitations under the License.
-//include { COPY_OUTPUT_TO_ENSEMBL_FTP as COPY_GENOME_OUTPUT } from '../modules/copy_output_to_ensembl_ftp.nf'
-//include { COPY_OUTPUT_TO_ENSEMBL_FTP as COPY_PROTEIN_OUTPUT } from '../modules/copy_output_to_ensembl_ftp.nf'
+// include { COPY_OUTPUT_TO_ENSEMBL_FTP as COPY_GENOME_OUTPUT } from '../modules/copy_output_to_ensembl_ftp.nf'
+// include { COPY_OUTPUT_TO_ENSEMBL_FTP as COPY_PROTEIN_OUTPUT } from '../modules/copy_output_to_ensembl_ftp.nf'
```
```suggestion
```

JAlvarezJarreta commented on pipelines/nextflow/subworkflows/run_busco.nf:
```diff
@@ -44,49 +42,83 @@ include { BUSCO_CORE_METAKEYS as BUSCO_CORE_METAKEYS_GENOME } from '../modules/b
+        // Run BUSCO in genome mode
+        if (busco_mode.contains('genome')) {
+
+            output_typeG = "genome"
```
```suggestion
            def output_typeG = 'genome'
```

JAlvarezJarreta commented on pipelines/nextflow/subworkflows/run_busco.nf:
```diff
@@ -44,49 +42,83 @@ include { BUSCO_CORE_METAKEYS as BUSCO_CORE_METAKEYS_GENOME } from '../modules/b
+            // Copy BUSCO summary stats to ensembl FTP
+            // if (copyToFtp) {
+            //     COPY_GENOME_OUTPUT(buscoGenomeSummaryOutput)
+            // }
```
```suggestion
```

JAlvarezJarreta commented on pipelines/nextflow/subworkflows/run_busco.nf:
```diff
@@ -44,49 +42,83 @@ include { BUSCO_CORE_METAKEYS as BUSCO_CORE_METAKEYS_GENOME } from '../modules/b
+        // Run Busco in protein mode
+        if (busco_mode.contains('protein')) {
+
+            output_typeP = "protein"
```
```suggestion
            def output_typeP = 'protein'
```

JAlvarezJarreta commented on pipelines/nextflow/subworkflows/run_busco.nf:
```diff
@@ -44,49 +42,83 @@ include { BUSCO_CORE_METAKEYS as BUSCO_CORE_METAKEYS_GENOME } from '../modules/b
 
+            buscoProteinSummaryOutput = BUSCO_PROTEIN_OUTPUT(buscoProteinOutput, output_typeP)
 
+            // Copy BUSCO summary stats to ensembl FTP
+            // if (copyToFtp) {
+            //     COPY_PROTEIN_OUTPUT(buscoProteinSummaryOutput)
+            // }
```
```suggestion
```

JAlvarezJarreta commented on pipelines/nextflow/subworkflows/run_busco.nf:
```diff
@@ -44,49 +42,83 @@ include { BUSCO_CORE_METAKEYS as BUSCO_CORE_METAKEYS_GENOME } from '../modules/b
+            // }
 
+            // Make and apply busco summary meta_keys patch directly to core:
+            if(params.apply_busco_metakeys){
```
```suggestion
            if (params.apply_busco_metakeys) {
```

JAlvarezJarreta commented on pipelines/nextflow/subworkflows/run_busco.nf:
```diff
@@ -44,49 +42,83 @@ include { BUSCO_CORE_METAKEYS as BUSCO_CORE_METAKEYS_GENOME } from '../modules/b
+            // }
+
+            // Make and apply busco summary meta_keys patch directly to core:
+            if(params.apply_busco_metakeys){
```
```suggestion
            if (params.apply_busco_metakeys) {
```

JAlvarezJarreta commented on pipelines/nextflow/subworkflows/run_ensembl_stats.nf:
```diff
@@ -38,21 +36,25 @@ include { POPULATE_DB as ADD_BETA_UPDATES_ON_CORE  } from '../modules/ensembl_st
+
+            statisticsFile = RUN_STATISTICS(db_meta)
+
+            if( params.apply_ensembl_stats ) {
```
```suggestion
            if ( params.apply_ensembl_stats ) {
```

JAlvarezJarreta commented on pipelines/nextflow/subworkflows/run_ensembl_stats.nf:
```diff
@@ -38,21 +36,25 @@ include { POPULATE_DB as ADD_BETA_UPDATES_ON_CORE  } from '../modules/ensembl_st
+
+            betaMetakeys = RUN_BETA_METAKEYS (db_meta)
+
+            if(params.apply_ensembl_beta_metakeys){
+                ADD_BETA_UPDATES_ON_CORE(betaMetakeys)
+                }
```
```suggestion
            if (params.apply_ensembl_beta_metakeys) {
                ADD_BETA_UPDATES_ON_CORE(betaMetakeys)
            }
```

JAlvarezJarreta commented on pipelines/nextflow/subworkflows/run_ensembl_stats.nf:
```diff
@@ -38,21 +36,25 @@ include { POPULATE_DB as ADD_BETA_UPDATES_ON_CORE  } from '../modules/ensembl_st
-            ADD_BETA_UPDATES_ON_CORE(betaMetakeys)
-            }
+
+            betaMetakeys = RUN_BETA_METAKEYS (db_meta)
```
```suggestion
            betaMetakeys = RUN_BETA_METAKEYS(db_meta)
```

ens-LCampbell replied:
Ill fix it in upcoming commit thanks

JAlvarezJarreta commented on pipelines/nextflow/subworkflows/run_omark.nf:
```diff
@@ -40,25 +37,33 @@ include { OMARK_OUTPUT } from '../modules/omark/omark_output.nf'
-        // 
-        def proteinData = FETCH_PROTEINS (db_meta.flatten())
-        //
+        proteinData = FETCH_PROTEINS (amended_db_meta)
```
```suggestion
        proteinData = FETCH_PROTEINS(amended_db_meta)
```

ens-LCampbell replied:
thanks good spot 👍

JAlvarezJarreta commented on pipelines/nextflow/subworkflows/run_omark.nf:
```diff
@@ -40,25 +37,33 @@ include { OMARK_OUTPUT } from '../modules/omark/omark_output.nf'
-        //if (params.copyToFtp) {
-        //    COPY_OMARK_OUTPUT(omarkSummaryOutput)
-        //}
+        // Copy Omark output to ensembl FTP
+        // if ( params.copyToFtp ) {
+        //     COPY_OMARK_OUTPUT(omarkSummaryOutput)
+        // }
```
```suggestion
```

JAlvarezJarreta commented on pipelines/nextflow/workflows/main.nf:
```diff
@@ -0,0 +1,97 @@
+
+// Some initialisation of default params
+if (params.busco_mode instanceof java.lang.String) {
+    busco_mode = params.busco_mode.split(/,/).collect().unique()
```
```suggestion
    def busco_mode = params.busco_mode.split(/,/).collect().unique()
```

JAlvarezJarreta commented on pipelines/nextflow/workflows/main.nf:
```diff
@@ -0,0 +1,97 @@
+    busco_mode = params.busco_mode.split(/,/).collect().unique()
+}
+else {
+    busco_mode = params.busco_mode
```
```suggestion
    def busco_mode = params.busco_mode
```

JAlvarezJarreta commented on pipelines/nextflow/workflows/main.nf:
```diff
@@ -0,0 +1,97 @@
+    validateParameters()
+    log.info paramsSummaryLog(workflow)
+
+    if(params.run_busco_ncbi && !params.run_busco_core){
```
```suggestion
    if (params.run_busco_ncbi && !params.run_busco_core) {
```

JAlvarezJarreta commented on pipelines/nextflow/workflows/main.nf:
```diff
@@ -0,0 +1,97 @@
+                    production_name:'dummy_prodname', organism_name:'DummyGenus dummyspecies', annotation_source:'dummy_anno_source']}
+                    .set { genome_metadata }
+
+            busco_mode = 'genome'
+            copyToFtp = false
```
```suggestion
            def busco_mode = 'genome'
            def copyToFtp = false
```

JAlvarezJarreta commented on pipelines/nextflow/workflows/main.nf:
```diff
@@ -0,0 +1,97 @@
+            RUN_BUSCO(genome_metadata, params.busco_mode, params.copyToFtp)
+        }
+        if (params.run_omark) {
+        RUN_OMARK(genome_metadata)
```
```suggestion
            RUN_OMARK(genome_metadata)
```

JAlvarezJarreta commented on pipelines/nextflow/workflows/main.nf:
```diff
@@ -0,0 +1,97 @@
+        RUN_OMARK(genome_metadata)
+        }
+        if (params.run_ensembl_stats || params.run_ensembl_beta_metakeys) {
+                RUN_ENSEMBL_STATS(genome_metadata)
```
```suggestion
            RUN_ENSEMBL_STATS(genome_metadata)
```

JAlvarezJarreta reviewed (changes requested):
A few minor indentation/old code left as comment/missing spaces.

Besides that, I have added the `def` keyword in every variable as this seems now to be the standard in Nextflow (in the past it looked like they were aiming to remove this to get Nextflow away from Groovy, but it is back 😢).

ens-ftricomi commented on pipelines/nextflow/modules/busco/busco_core_metakeys.nf:
```diff
@@ -16,47 +15,23 @@ See the License for the specific language governing permissions and
     input:
-    tuple val(gca), val(dbname),path(summary_file)
-    
+        tuple val(insdc_acc), val(dbname), val(formated_sci_name),
```
what is formated_sci_name?

ens-LCampbell replied:
This is the formatted organism name from BUSCO_OUTPUT.
`def formated_sci_name = organism_name.replaceAll("\\s", "_")`

ens-ftricomi commented on pipelines/nextflow/modules/busco/busco_dataset.nf:
```diff
@@ -17,23 +16,26 @@ limitations under the License.
     label 'python'
-    tag "$taxon_id:$dbname"
+    conda "${projectDir}/bin/environment.yml"
+    tag "${organism_name} [tax_id:${taxonomy_id}]"
```
if multiple gca for the same species i.e humans are running at the same time, the tag is not indicative anymore, I would add the db name anyway

JAlvarezJarreta replied:
Would something like `tag "${organism_name} [${insdc_acc}]"` also work for your needs @ens-ftricomi? This will also make the code more future proof for when the databases go away.

ens-LCampbell replied:
I will wait for feedback on this 👍

ens-ftricomi commented on pipelines/nextflow/modules/busco/busco_dataset.nf:
```diff
@@ -17,23 +16,26 @@ limitations under the License.
+        tuple val(insdc_acc), val(taxonomy_id), val(dbname),
+            val(production_name), val(organism_name), val(annotation_source),
+            env(orthodb), emit: clade_dataset
+        path("orthodb_set.txt")
```
if the output of the python script is an enviromental variable, why do we need a txt?
which is the behaviour of environmental variables when multiple dbs are processed in parallel ?

ens-LCampbell replied:
I was just keeping to what was implemented originally (i.e. previous: `stdout`). The `env()` output is the same, but we also retain a record in the orthodb_set.txt with no overhead.
As for parallel processing this isn't an issue as the meta_map in this case includes all linked info via a tuple, to ensure linking of genome+ortho for all downstream processing i.e.  block potential mixing of meta_map and orthodb_clade sets in parallel processing.

ens-ftricomi commented on pipelines/nextflow/modules/busco/busco_dataset.nf:
```diff
@@ -17,23 +16,26 @@ limitations under the License.
-    """
-
-    
+        output = 'orthodb_set.txt'
```
see above

ens-ftricomi commented on pipelines/nextflow/modules/busco/busco_dataset.nf:
```diff
@@ -17,23 +16,26 @@ limitations under the License.
+        orthodb = ''
+        """
+        clade_selector.py -d $params.busco_datasets_file -t $taxonomy_id > $output
+        orthodb=`cat ${output}`
```
see above

ens-ftricomi commented on pipelines/nextflow/modules/busco/busco_output.nf:
```diff
@@ -16,48 +15,41 @@ See the License for the specific language governing permissions and
-    """
-    sed '/Summarized benchmarking in BUSCO notation for file/d' $summary_file > $summary_name
-    """
+        formated_sci_name = organism_name.replaceAll("\\s", "_")
```
is there a default value for organism_name if it is missing from the meta table?

ens-LCampbell replied:
No there's no default. In this case it will be flagged as warning. I think all cores required to have `species.scientific_name`

ens-ftricomi commented on pipelines/nextflow/modules/busco/busco_output.nf:
```diff
@@ -16,48 +15,41 @@ See the License for the specific language governing permissions and
-    sed '/Summarized benchmarking in BUSCO notation for file/d' $summary_file > $summary_name
-    """
+        formated_sci_name = organism_name.replaceAll("\\s", "_")
+        publish_dir_name = formated_sci_name + '/' + insdc_acc + '/' + annotation_source
```
this structure is no longer needed as we don't need to sync to the ftp anymore, it can just be the GCA

ens-ftricomi commented on pipelines/nextflow/modules/cleaning.nf:
```diff
@@ -17,16 +17,16 @@ limitations under the License.
 
```
The module is not used, review not needed

ens-LCampbell replied:
No worries. Given it was present in the original branch, I didn't want to just delete it. Would you prefer it be dropped ?

ens-ftricomi commented on pipelines/nextflow/modules/copy_output_to_ensembl_ftp.nf:
```diff
@@ -1,4 +1,3 @@
-#!/usr/bin/env nextflow
```
module not needed anymore, review not needed

JackCurragh replied:
Does it make sense to update this PR to drop this functionality or merge the PR maintaining the same funcitonality and then do a new PR to drop it? 

I guess we don't want to merge things that we do not actually want so it might be better to cut it now?

ens-LCampbell replied:
Happy to drop it, wasn't sure at PR time given originally it wasn't omitted and seemed to still be in use in some cases.

ens-ftricomi commented on pipelines/nextflow/modules/ensembl_statistics/run_ensembl_meta.nf:
```diff
@@ -16,48 +15,36 @@ See the License for the specific language governing permissions and
     
+    script:
+        formated_sci_name = organism_name.replaceAll("\\s", "_")
+        publish_dir_name = formated_sci_name + '/' + insdc_acc + '/' + annotation_source
```
same comment I did for busco the directory structure can be easier so no need to create the full path, it can be the GCA so we can avoid to query the meta table

ens-ftricomi commented on pipelines/nextflow/modules/ensembl_statistics/run_statistics.nf:
```diff
@@ -16,39 +15,32 @@ See the License for the specific language governing permissions and
-    
-
+        formated_sci_name = organism_name.replaceAll("\\s", "_")
+        publish_dir_name = formated_sci_name + '/' + insdc_acc + '/' + annotation_source
```
see above

ens-ftricomi commented on pipelines/nextflow/modules/fetch_genome_core.nf:
```diff
@@ -16,33 +15,33 @@ See the License for the specific language governing permissions and
+    genome_fasta = "genome_toplevel.fa"
+    sequence_dumping_script = file("${params.enscode}/ensembl-analysis/scripts/sequence_dump.pl")
+
+    """
```
sorry which is the difference here? just indendation?

JackCurragh replied:
And change to where the script path is defined.

ens-ftricomi commented on pipelines/nextflow/modules/omark/omamer_hog.nf:
```diff
@@ -17,18 +16,22 @@ limitations under the License.
-    omamer search --db ${params.omamer_database} --query ${translation_file} --score sensitive --out proteins.omamer 
-    """
+        """
+        omamer search --db ${params.omamer_database} --query ${aa_or_genome_seqs} --score sensitive --out proteins.omamer 
```
the input is just translation file so I would leave the original variable name

ens-LCampbell replied:
Happy to revert for Omark. I maintained the same as in modules/busco/busco_lineages.nf for better linkage across the two modules, given modules/busco/busco_lineages.nf it can be Aa or Nuc

ens-ftricomi commented on pipelines/nextflow/modules/omark/omark.nf:
```diff
@@ -16,28 +15,24 @@ See the License for the specific language governing permissions and
+
     output:
-    tuple val(gca), val(db), val(publish_dir), path("omark_output/*.txt"), path("omark_output/*")
+        tuple val(insdc_acc), val(dbname), val(formated_sci_name), val(publish_dir_name), path("omark_output/*_summary.txt"), path("omark_output/*")
```
I think the omark output doesn't end for summary.txt, did you check?

ens-LCampbell replied:
From the output on tests its coming out a `proteins_detailed_summary.txt`

ens-ftricomi commented on pipelines/nextflow/modules/omark/omark_output.nf:
```diff
@@ -16,30 +15,23 @@ See the License for the specific language governing permissions and
-    cat  $summary_file >  $summary_name
-    
-    """
+        accession_formatted = insdc_acc.toLowerCase().replaceAll(/\./, "v").replaceAll(/_/, "")
```
same comment I made before the output ca be as it is so this logic is no longer needed

ens-ftricomi commented on pipelines/nextflow/modules/query_core_meta.nf:
```diff
@@ -0,0 +1,34 @@
+/*
+See the NOTICE file distributed with this work for additional information
```
general comment on this module: all the metakeys were used before to build the folder structure to make our life easier in order to sync the results into the FTP. As this is no longer needed we can avoid this step and the only thing we can use is the core db or the GCA that can be retrieved from the db name. So we can avoid calling genomio and anyway I think this  task can be moved to a simply python script if we still want to retrieve the metakeys
So I would skip the module or add a python script for this in the bin folder

JAlvarezJarreta replied:
Not every core database has the right accession in their name (e.g. GCF are not allowed in main sites), so I'm afraid the solution you are proposing does not work for all our core databases at the minute (once main sites are frozen and we are only importing new things for Beta that will be different).

If you want to make the pipeline completely disconnected from the core dbs, we could decide to use something like datasets CLI to fetch the assembly FASTA file from GenBank giving just the accession and scientific name in the input file. But I would argue that is out of the scope of this PR.

ens-ftricomi commented on pipelines/nextflow/subworkflows/prepare_metadata.nf:
```diff
@@ -0,0 +1,50 @@
+        core_metakeys // input meta keys JSON
+
+    main:
+        QUERY_CORE_META(input_cores, core_metakeys)
```
why the output of QUERY_CORE_METADATA needs an extra step? Can the output be emitted directly by the module itself?

ens-LCampbell replied:
The setup of this was originally a module, but I've opted for a subworkflow as the combined module processing of `QUERY_CORE_META` followed by a function `index_metadata` to map the output meta into a named meta_map is much more aligned to subworkflow than a module alone.

ens-ftricomi commented on pipelines/nextflow/subworkflows/run_ensembl_stats.nf:
```diff
@@ -38,21 +36,25 @@ include { POPULATE_DB as ADD_BETA_UPDATES_ON_CORE  } from '../modules/ensembl_st
-            ADD_STATS_ON_CORE(statisticsFile)
+        if( params.run_ensembl_stats ) {
+
+            statisticsFile = RUN_STATISTICS(db_meta)
```
def statisticsFile

ens-LCampbell replied:
Ill add it sure

ens-ftricomi commented on pipelines/nextflow/subworkflows/run_omark.nf:
```diff
@@ -18,19 +17,17 @@ limitations under the License.
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 */
 
+include { COPY_OUTPUT_TO_ENSEMBL_FTP as COPY_OMARK_OUTPUT } from '../modules/copy_output_to_ensembl_ftp.nf'
```
no needed

ens-ftricomi commented on pipelines/nextflow/workflows/nextflow.config:
```diff
@@ -1,32 +1,56 @@
 
 // Set minimum required Nextflow version, stopping execution if current version does not match
-//nextflowVersion = '!>=23.01'
+nextflowVersion = '!>=23.10'
```
if it works with the default version in the cluster I would relax this condition

ens-LCampbell replied:
The default we have (as standard after logging into codon) 
`nextflow -version` -> `version 24.10.3`
Also, given we want to use minimum nf-schema version 2.2.0 which requires nextflow version 23.10+

ens-ftricomi commented on pipelines/nextflow/workflows/nextflow.config:
```diff
@@ -38,52 +62,48 @@ params {
 
-  jdbcUrl = "jdbc:mysql://${params.host}:${params.port}/${params.dbname}"
+  ncbiBaseUrl = "https://api.ncbi.nlm.nih.gov/datasets/v2alpha/genome/accession/"
+  production_ftp_dir = "/nfs/production/flicek/ensembl/production/ensemblftp/rapid-release/species/"
```
no needed

ens-ftricomi commented on pipelines/nextflow/subworkflows/run_busco.nf:
```diff
@@ -44,49 +42,83 @@ include { BUSCO_CORE_METAKEYS as BUSCO_CORE_METAKEYS_GENOME } from '../modules/b
+            BUSCO_PROTEIN_LINEAGE(metadata_prot_trans, output_typeP).busco_report_output
+                .map { row -> [
+                    insdc_acc: row[0], taxonomy_id: row[1], core: row[2], production_name: row[3],
+                    organism_name: row[4], annotation_source: row[5], report: row[6]
```
what is the report variable?

ens-LCampbell replied:
Just the `busco_report_output` summary report generated within the BUSCO_LINEAGES module

ens-LCampbell committed:
Apply suggestions from code review @JorgeAlvarez

Code edits/tweaks. Particularly re introduction of def variable declarations

Co-authored-by: Jorge Alvarez-Jarreta <jalvarez@ebi.ac.uk>

ens-LCampbell committed:
Update pipelines/nextflow/modules/omark/omark.nf

Fix module - setup order

Co-authored-by: Jorge Alvarez-Jarreta <jalvarez@ebi.ac.uk>

ens-LCampbell commented:
> A few minor indentation/old code left as comment/missing spaces.
> 
> Besides that, I have added the `def` keyword in every variable as this seems now to be the standard in Nextflow (in the past it looked like they were aiming to remove this to get Nextflow away from Groovy, but it is back 😢).

Fixed. Thanks Jorge, that couldn't of been a fun task. Thanks to hackathon we realised, but 👎 to nextflow flip flop

JackCurragh commented on pipelines/nextflow/workflows/nextflow_schema.json:
```diff
@@ -0,0 +1,364 @@
+{
+  "$schema": "https://json-schema.org/draft/2020-12/schema",
+  "$id": "https://raw.githubusercontent.com/nf-core/testpipeline/master/nextflow_schema.json",
```
This should point to the url of this file rather than the template.

ens-LCampbell replied:
Ahh, alrighty I didn't know that. Wouldn't I need to wait until merged before updating it to URL on the repo ?

JackCurragh replied:
Yeah I always struggled with this circularness with nf-core. You need to merge the PR to make the link and then make a new PR to add the link. Very annoying.

So I guess this url is fine here but would need a second PR soon after to fix it. I am not all that clear on what the URL is used for or what it being wrong would break.

JackCurragh commented on pipelines/nextflow/workflows/nextflow_schema.json:
```diff
@@ -0,0 +1,364 @@
+        },
+        "team": {
+          "type": "string",
+          "description": "Required CoreDB meta_key for data team responsible.",
```
I am not familiar with the CoreDB meta_keys but it occurs to me that this may be more appropriate as an enum to avoid typo related issues down the line?

JackCurragh replied:
Just realised this is set as an ENV variable so likely won't be misspelled but still might be wise to update but probably fine if not done

ens-LCampbell replied:
Pros and Cons I guess. If team names change, expand or reduce this would need updating. Happy to add it though, lets see how it plays out

ens-LCampbell replied:
"enum": ["genebuild","metazoa","plants","microbes"] Will be added who have I missed hmm

JackCurragh commented on pipelines/nextflow/workflows/nextflow_schema.json:
```diff
@@ -0,0 +1,364 @@
+          "fa_icon": "fas fa-folder-open",
+          "hidden": true
+        },
+        "copyToFtp": {
```
@ens-ftricomi can this param be cut if we are not supporting copying to FTP anymore?

ens-LCampbell replied:
I think it needs dropped yeah

JackCurragh commented on pipelines/nextflow/workflows/nextflow.config:
```diff
@@ -147,24 +176,33 @@ process {
+    time = { 0.2.h * task.attempt }
+    module = 'singularity-3.7.0-gcc-9.3.0-dp5ffrp'
+    container = "ensemblorg/ensembl-genomio:${params.genomio_version}"
+    afterScript = { "sleep 30" }
```
Is there a reason $params.files_latency isnt used here?

ens-LCampbell replied:
Its too high, so I opted to hard code it for genomio. I played with this setting and 30 is plenty. Note, sleep definitely still needed just 60s default is too much. Also, I didn't feel like bloating schema for one modules afterScript

JackCurragh commented on pipelines/nextflow/workflows/nextflow_schema.json:
```diff
@@ -0,0 +1,364 @@
+          "fa_icon": "fas fa-cog",
+          "hidden": true
+        },
+        "download_path": {
```
duplicate of L203

JackCurragh replied:
I would actually advise removing the one on L203 to keep BUSCO stuff together

ens-LCampbell replied:
Yep Agreed ! Now gone 👍

JackCurragh commented on pipelines/nextflow/workflows/nextflow_schema.json:
```diff
@@ -0,0 +1,364 @@
+        }
+      },
+      "dependentRequired": {
+        "bioperl": ["enscode"]
```
I think this is backwards. "enscode" is a required dependency of "bioperl" 

Change to 
```
      "dependentRequired": {
        "enscode": ["bioperl"]
      }
```

JackCurragh commented on pipelines/nextflow/workflows/nextflow_schema.json:
```diff
@@ -0,0 +1,364 @@
+          "required": ["run_busco_ncbi"]
+        }
+      },
+      "dependentRequired": {
```
Again, I _believe_ the logic here is backwards. Was this checked with the nf-core tooling? I think it should throw errors in these cases. 

Replace with: 
```
      "dependentRequired": {
        "team": ["run_ensembl_beta_metakeys"],
        "host": ["run_busco_core", "run_omark"],
        "port": ["run_busco_core", "run_omark"],
        "password": ["run_busco_core", "run_omark"],
        "server_set": ["run_busco_core", "run_omark"]
      }
      ```

ens-LCampbell replied:
I tested this and it doesn't work. The logic as it is now is correct. I did remove some of the preceeding if, then logic though thank you for checking the schema. Its not fully valid and compliant (checked via nf-core v3.2.0)

ens-LCampbell replied:
An update in later versions of nf-core, validation recommends to remove all default values in the schema and only retain those in nextflow.config. All defaults now dropped.

JackCurragh replied:
Ah interesting. I check this with nf-core rather than actually running the pipeline and nf-core complained with the current setup and not with my suggestion. Obviously the pipeline actually running is a more important metric.

JackCurragh commented on pipelines/nextflow/workflows/nextflow_schema.json:
```diff
@@ -0,0 +1,364 @@
+        "run_busco_core": {
+          "type": "boolean",
+          "description": "Run BUSCO given a mysql db.",
+          "default": false,
```
I think setting defaults here and in the nextflow.config is a bad idea. I think anything set here would be superseded by what is in the nextflow.config anyway so these defaults can be dropped (?)

ens-LCampbell replied:
Ahh, good observation. I made the same mistake, in fact this schema "default" isn't doing anything other than printing the actual default (nf config) to the user. Its not a required key, and would need updating if the config default changes

JackCurragh commented on pipelines/nextflow/workflows/nextflow_schema.json:
```diff
@@ -0,0 +1,364 @@
+        "files_latency": {
+          "type": "integer",
+          "format": "[0-9]+",
+          "default": "60",
```
Should be int not str right?

ens-LCampbell replied:
Indeed ! Thanks for the spot

JackCurragh commented on pipelines/nextflow/workflows/nextflow_schema.json:
```diff
@@ -0,0 +1,364 @@
+          "items": {
+            "enum": ["protein", "genome", "genome,protein", "protein,genome"]
+          },
+          "default": ["protein", "genome"],
```
Default type does not match property type. change to "protein,genome"?

ens-LCampbell replied:
Yep, thanks. Mode stuff was changed recently following testing from @EreboPSilva 👍

JackCurragh commented on pipelines/nextflow/workflows/nextflow.config:
```diff
@@ -93,6 +113,10 @@ singularity {
 
 profiles {
 
+  conda {
```
I recall during my placement almost 2 years ago being told that conda wasn't suitable for the cluster. I guess this has changed. Is using conda in Nextflow okay as it just builds the envs in the workdir and wont lead to bloat longterm?

ens-LCampbell replied:
Theres the option of using a shared cache space, much in the same as NXF_SINGULARITY_CACHEDIR. So once an environment is created, it can be shared among a team. In plantazoas case we have a shared NXF_CODON_CACHEDIR on nfs. 

From nxf docs:
`Directory where Conda environments are stored. When using a computing cluster it must be a shared folder accessible from all compute nodes.`

JackCurragh commented on pipelines/nextflow/workflows/main.nf:
```diff
@@ -0,0 +1,97 @@
+}
+else {
+    def busco_mode = params.busco_mode
+}
```
Is this needed? looks lke busco_mode is defined when used below. 

Also believe java.lang is not needed and `instanceof String` is valid

JackCurragh commented on pipelines/nextflow/workflows/input_csv_schema.json:
```diff
@@ -0,0 +1,39 @@
+        "items": {
+            "type": "object",
+            "properties": {
+                "core": {
```
Should core instead be 'database_name'? That is how it is accessed in main.nf (`row.database_name`)

ens-LCampbell replied:
Makes sense sure. Though I ensured to maintain old style to not change basic assumptions. @ens-ftricomi Is this alright to update ?

JackCurragh commented on pipelines/nextflow/modules/busco/busco_dataset.nf:
```diff
@@ -17,23 +16,26 @@ limitations under the License.
+        def output = 'orthodb_set.txt'
+        def orthodb = ''
+        """
+        clade_selector.py -d $params.busco_datasets_file -t $taxonomy_id > $output
```
Calling params.anything within a module makes the module less portable and as a result is against DSL 2 conventions. We should move all $params calls to workflow level and keep the modules clean.

JAlvarezJarreta replied:
Whilst I agree it would make it more portable (and should be done if the module is put in a central modules repo), I would like to know more about the DSL2 part, as in the Nextflow documentation seems to be "allowed": https://www.nextflow.io/docs/latest/dsl1.html#processes-and-workflows (I find that sometimes they say separate things in their documentation in different places, so if you have more info, I would love to read it)

JackCurragh replied:
I think you are right and conventions is a bit too strong a term. Perhaps DSL2 'ethos' of modularity. 

Definitely not a required change (as is the case for most of my comments) I was just commenting as I parsed through it all.

JackCurragh commented on pipelines/nextflow/workflows/bin/environment.yml:
Assuming conda is allowed (I mention my confusiton in another comment), perhaps we can devise a policy for storing these files. `bin` is for executables rather than config files. `workflows/conda/busco-omark-stats.yml` might work?

ens-LCampbell replied:
Happy to follow any approach once agreed 👍 Can I ask you to raise this and get back to me ? Your suggestion seems reasonable to me !

JackCurragh commented on pipelines/nextflow/workflows/main.nf:
```diff
@@ -0,0 +1,97 @@
+            Channel.fromList(samplesheetToList(params.csvFile, file("${projectDir}/input_csv_schema.json")))
+                    .flatten()
+                    .map { row -> [insdc_acc:row.accession, taxonomy_id:row.taxon_id, core:'dummy_coredb', \
+                    production_name:'dummy_prodname', organism_name:'DummyGenus dummyspecies', annotation_source:'dummy_anno_source']}
```
I can't see where these 'dummy' entries get set to real values? 

This might just be me struggling as it is a big PR but if busco_mode is 'protein' the 'dummy_prodname' will be used in the translations file and still contain the dummy variable?

ens-LCampbell replied:
To my understanding, these dummy variables are not vital more for feedback to user when parallel processing multiple cores/genomes. In this case, its non-db GCA based processing so only a subset of meta_map values can be applied. So dummy variables (all the same) used in the remainder of the meta_map values.

This will go away in the following PR to update (I alluded to in the meta_map formatting question above). 

so module input: would be 
`tuple val(meta), val(x), val(y)`
as opposed to current:
`tuple val(insdc_acc), val(taxonomy_id), val(dbname),....`

JackCurragh commented on pipelines/nextflow/subworkflows/run_busco.nf:
```diff
@@ -44,49 +40,75 @@ include { BUSCO_CORE_METAKEYS as BUSCO_CORE_METAKEYS_GENOME } from '../modules/b
-            BUSCO_CORE_METAKEYS_PROTEIN(buscoProteinSummaryOutput)
+        // Get the closest BUSCO dataset from the taxonomy classification stored in db meta table
+        BUSCO_DATASET(db_meta).clade_dataset
+            .map { row -> [
```
Does mapping outputs every time have any advantages? Personally I feel it complicates things.

JackCurragh replied:
I get that it follows the idea of 'Code as documentation' but it kind of looks messy in this layout. 

Would a comment with the tuple structure work? Or maybe format it like this? 
```
BUSCO_PROTEIN_LINEAGE(metadata_prot_trans, output_typeP).busco_report_output
                .map { 
                    row -> [
                        insdc_acc: row[0],
                        taxonomy_id: row[1],
                        core: row[2],
                        production_name: row[3],
                        organism_name: row[4],
                        annotation_source: row[5],
                        report: row[6]
                    ]
                }
                .set{ buscoProteinOutput }
 ```

ens-LCampbell replied:
So this is messy 100% agree, we're actually just updating our own thinking (some discussion and insight from a recent nextflow hackathon) on meta_maps and how they should be defined, supplied and expanded within modules. Some old methods persisting here, whereas actually just passing the input as a single meta_map, and using the meta_map.some_index within the module would make it more flexible. As you might see, the base meta_map (6 values) requires updating along the workflow, which is tricker than one might expect 😓  

Anyway, I think why not ! It does read better so happy to reformat the structure. My plan is to address the meta_map in a following PR to make it less verbose.

JackCurragh commented on pipelines/nextflow/modules/busco/busco_lineages.nf:
```diff
@@ -0,0 +1,60 @@
+process BUSCO_LINEAGES {
+    label 'busco'
+    tag "${organism_name}:${insdc_acc}"
+    storeDir "${params.cacheDir}/${insdc_acc}/busco_${busco_mode}/"
```
I cannot believe I didn't know about this directive! I have devised some awfully convoluted methods to mimic this and none of them have worked all that well.

JackCurragh commented on pipelines/nextflow/modules/fetch_genome_core.nf:
```diff
@@ -16,33 +15,33 @@ See the License for the specific language governing permissions and
+
   input:
-  tuple val(gca), val(dbname), val(busco_dataset)
+    tuple val(gca), val(dbname), val(busco_dataset)
```
Is GCA correct? Seems like in other places an effort has been made to call this `insdc_acc`.

ens-LCampbell replied:
This module isn't used in the current iteration of the pipeline, but was an additional module so using old input var names. @ens-ftricomi Is this ok to drop or do you have plans for it ?

JackCurragh reviewed (commented):
Due to my relative newness my review is quite superficial. It was definitely valuable for me and hopefully some of my comments can be constructive. 

There are aspects that are considerable improvements but I think it will be hard to merge this particular PR.

Commit:
PR review update config and nxf schema

Commit:
Update+tested modules/subwfs PR review

Commit:
Remove brc from allowed project

ens-LCampbell committed:
Merge pull request #1 from Ensembl/main

Update licensing to pep-0639

Commit:
Update path to generate_species_homepage_stats

leannehaggerty commented:
After discussion within the development team, we've agreed that the current main branch is not the appropriate target for this pipeline. This repository is intended to support annotation-focused Ensembl Nextflow pipelines and will be restructured to provide reusable modules and workflows. As part of this restructuring, the existing main branch will be treated as a developmental branch.

Given this direction, it is not suitable to merge this pull request into main at this time. Instead, teams should continue working from whichever branch best suits their current needs.

Additionally, we encourage teams to consider integrating the functionality of the stats and metadata scripts into their pipelines where helpful. However, please note that these are temporary tools to support the beta phase and are not intended for long-term use.
